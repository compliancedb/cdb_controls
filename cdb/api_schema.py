class ApiSchema:
    @staticmethod
    def url_for_owner_projects(host, project_data):
        return host + '/api/v1/projects/' + project_data["owner"] + "/"

    @staticmethod
    def url_for_project(host, project_data):
        return ApiSchema.url_for_owner_projects(host, project_data) + project_data["name"]

    @staticmethod
    def url_for_releases(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/releases/'

    @staticmethod
    def url_for_commit(host, project_data, commit):
        return ApiSchema.url_for_project(host, project_data) + '/commits/' + commit

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_artifact(host, project_data, sha256_digest):
        return ApiSchema.url_for_artifacts(host, project_data) + sha256_digest

    @staticmethod
    def url_for_artifacts(host, project_data):
        return ApiSchema.url_for_project(host, project_data) + '/artifacts/'

    @staticmethod
    def url_for_release(host, project_data, release_number):
        return ApiSchema.url_for_releases(host, project_data) + release_number