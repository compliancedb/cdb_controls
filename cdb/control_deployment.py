def control_deployment_approved(approvals):
    for approval in approvals:
        if approval["state"] == "APPROVED":
            return True
    return False
