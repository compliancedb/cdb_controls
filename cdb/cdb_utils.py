import getopt
import json
import os
import sys
import docker
import requests as req
from junitparser import JUnitXml
from requests.auth import HTTPBasicAuth

from cdb.settings import CDB_SERVER

CMD_HELP = 'ensure_project.py -p <project.json>'


def project_exists_in_cdb(project_data, projects):
    exists = False
    for project in projects:
        if project["name"] == project_data["name"]:
            exists = True
    return exists


def get_project_list_from_cdb(projects_url):
    resp = req.get(projects_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    projects = resp.json()
    return projects


def get_project_from_cdb(project_url):
    resp = req.get(project_url)
    # If the response was successful, no Exception will be raised
    resp.raise_for_status()
    print(resp.text)
    project = resp.json()
    return project


def load_project_configuration(json_data_file):
    project_data = json.load(json_data_file)
    return project_data


def create_artifact(api_token, host, project_config_file, sha256, filename, description, git_commit, commit_url, build_url,
                    is_compliant):
    project_data = load_project_configuration(project_config_file)

    '''
    curl -H 'Content-Type: application/json' \
     -X PUT \
     -d '{"sha256": "'"$3"'", "filename": "'"$4"'", "description": "'"$5"'", "git_commit": "'"$6"'", "commit_url": "'"$7"'", "build_url": "'"$8"'", "is_compliant": "true"}' \
    http://server:8001/api/v1/projects/$1/$2/artifacts/
    '''

    create_artifact_payload = {
        "sha256": sha256,
        "filename": filename,
        "description": description,
        "git_commit": git_commit,
        "commit_url": commit_url,
        "build_url": build_url,
        "is_compliant": is_compliant
    }
    url = url_for_artifacts(host, project_data)
    put_payload(create_artifact_payload, url, api_token)


def put_payload(payload, url, api_token):
    headers = {"Content-Type": "application/json"}
    print("Putting this payload:")
    print(json.dumps(payload, sort_keys=True, indent=4))
    print("To url: " + url)
    resp = req.put(url, data=json.dumps(payload), headers=headers, auth=HTTPBasicAuth(api_token, 'unused'))
    print(resp.text)


def url_for_artifacts(host, project_data):
    return url_for_project(host, project_data) + project_data["name"] + '/artifacts/'


def add_evidence(api_token, host, project_file_contents, sha256_digest, evidence):
    project_data = load_project_configuration(project_file_contents)
    url_for_artifact = url_for_artifacts(host, project_data) + sha256_digest

    put_payload(evidence, url_for_artifact, api_token)


def url_for_project(host, project_data):
    return host + '/api/v1/projects/' + project_data["owner"] + '/'


def rchop(thestring, ending):
    if thestring.endswith(ending):
        return thestring[:-len(ending)]
    return thestring


def parse_cmd_line():
    project_file = ""
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hp:", ["project="])
    except getopt.GetoptError:
        print(CMD_HELP)
        sys.exit(2)
    for opt, arg in opts:
        print(opt + ": " + arg)
        if opt == '-h':
            print(CMD_HELP)
            sys.exit()
        elif opt in ("-p", "--project"):
            project_file = arg
    return project_file


def get_image_details():
    docker_image = os.getenv('CDB_DOCKER_IMAGE', "NO_DOCKER_IMAGE_FOUND")
    sha256_digest = os.getenv('CDB_ARTIFACT_SHA', None)
    if sha256_digest is None:
        client = docker.from_env()
        print("Inspecting docker image for sha256Digest")
        image = client.images.get(docker_image)
        sha256_digest = image.attrs["RepoDigests"][0].split(":")[1]
    else:
        print("Docker image digest found from environment variable")

    print("IMAGE DIGEST: " + str(image.attrs["RepoDigests"][0]))
    return docker_image, sha256_digest


def env_is_compliant():
    return os.getenv('CDB_IS_COMPLIANT', "FALSE") == "TRUE"


def is_compliant_suite(junit_xml):
    if junit_xml.failures != 0:
        return False, "Tests contain failures"
    if junit_xml.errors != 0:
        return False, "Tests contain errors"
    return True, "All tests passed"


def load_test_results(file_path):
    test_xml = JUnitXml.fromfile(file_path)
    return test_xml


def is_compliant_test_results(file_path):
    """
    This parses a junit xml file to determine if there are any errors or failures

    return:
        A tuple, with is_compliant, plus a message string
    """
    test_xml = load_test_results(file_path)
    if test_xml._tag == "testsuites":
        for suite in test_xml:
            suite_is_compliant, message = is_compliant_suite(suite)
            if not suite_is_compliant:
                return suite_is_compliant, message
        # every test suite passed, so return True
        return True, "All tests passed"
    if test_xml._tag == "testsuite":
        return is_compliant_suite(test_xml)
    return False, "Could not find test suite(s)"


def control_junit():
    print("Publish evidence to ComplianceDB")
    junit_results_path = "/data/junit/junit.xml"

    is_compliant, message = is_compliant_test_results(junit_results_path)
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "junit")
    description = "JUnit results xml verified by compliancedb/cdb_controls: " + message
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def put_evidence():
    print("Publish evidence to ComplianceDB")

    is_compliant = env_is_compliant()
    evidence_type = os.getenv('CDB_EVIDENCE_TYPE', "EVIDENCE_TYPE_UNDEFINED")
    description = os.getenv('CDB_DESCRIPTION', "UNDEFINED")
    build_url = os.getenv('CDB_CI_BUILD_URL', "URL_UNDEFINED")

    evidence = build_evidence_dict(is_compliant, evidence_type, description, build_url)
    send_evidence(evidence)


def send_evidence(evidence):
    print(evidence)
    project_file = parse_cmd_line()
    with open(project_file) as project_file_contents:
        _docker_image_unused, sha256_digest = get_image_details()
        api_token = os.getenv('CDB_API_TOKEN', 'NO_API_TOKEN_DEFINED')
        host = os.getenv('CDB_HOST', CDB_SERVER)
        add_evidence(api_token, host, project_file_contents, sha256_digest, evidence)


def build_evidence_dict(is_compliant, evidence_type, description, build_url):
    evidence = {"evidence_type": evidence_type, "contents": {
        "is_compliant": is_compliant,
        "url": "",
        "description": ""
    }}
    evidence["contents"]["description"] = description
    evidence["contents"]["url"] = build_url
    return evidence
