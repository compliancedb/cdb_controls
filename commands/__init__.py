from .command_error import CommandError

from .json_loader import load_json

from .context import Context

from .command import Command

from .declare_pipeline import DeclarePipeline
from .control_deployment import ControlDeployment
from .log_artifact import LogArtifact
from .log_deployment import LogDeployment
from .log_evidence import LogEvidence
from .log_approval import LogApproval
from .log_test import LogTest

from .command_builder import build_command, COMMANDS
from .runner import run, main
