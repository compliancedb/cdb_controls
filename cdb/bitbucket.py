import json
import os
import sys

import requests

from cdb.cdb_utils import build_evidence_dict, send_evidence


def put_bitbucket_pull_request(project_config_file):

    pull_requests = get_pull_request_for_current_commit()

    print("Found pull request info: " + str(pull_requests))
    evidence = build_evidence_dict(
        is_compliant=True,
        evidence_type="pull_request",
        description="Bitbucket pull request",
        build_url=pull_requests[0]["pullRequestURL"])
    evidence["contents"]["source"] = pull_requests
    send_evidence(project_config_file, evidence)


def get_pull_request_for_current_commit():
    workspace = os.getenv('BITBUCKET_WORKSPACE', None)
    repository = os.getenv('BITBUCKET_REPO_SLUG', None)
    commit = os.getenv('BITBUCKET_COMMIT', None)
    user = os.getenv('BITBUCKET_USER', None)
    password = os.getenv('BITBUCKET_PWD', None)
    pull_requests = get_pull_requests_from_bitbucket_api(
        workspace=workspace, repository=repository, commit=commit, username=user, password=password)
    return pull_requests


def get_pull_requests_from_bitbucket_api(workspace, repository, commit, username, password):
    pull_requests_evidence = []

    url = f"https://api.bitbucket.org/2.0/repositories/{workspace}/{repository}/commit/{commit}/pullrequests"
    r = requests.get(url, auth=(username, password))
    if r.status_code == 200:
        pull_requests_json = json.loads(r.text)
        pull_requests = pull_requests_json["values"]
        for pr in pull_requests:
            pr_evidence = {}

            pr_html_url = pr['links']['html']['href']
            pr_api_url = pr['links']['self']['href']

            pr_evidence['pullRequestMergeCommit'] = commit
            pr_evidence['pullRequestURL'] = pr_html_url

            pr_evidence = get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password)

            pull_requests_evidence.append(pr_evidence)

    elif r.status_code == 202:
        print("Repository pull requests are still being indexed, please retry.")
        sys.exit(1)
    elif r.status_code == 404:
        print("Repository does not exists or pull requests are not indexed. Please make sure Pull Request Commit Links app is installed")
        sys.exit(2)
    else:
        print(f"Exception occured in fetching pull requests. Http return code is {r.status_code}")
        sys.exit(3)

    return pull_requests_evidence


def get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password):
    r = requests.get(pr_api_url, auth=(username, password))
    if r.status_code == 200:
        pr_json = json.loads(r.text)
        pr_evidence['pullRequestState'] = pr_json['state']
        participants = pr_json['participants']
        approvers = ""
        if len(participants) > 0:
            for participant in participants:
                if participant['approved']:
                    approvers = approvers + participant['user']['display_name'] + ","
            approvers = approvers[:-1]
            pr_evidence['approvers'] = approvers
        else:
            print("No approvers found")
    else:
        print("Error occured in fetching pull request details. Please review repository permissions.")
    return pr_evidence





