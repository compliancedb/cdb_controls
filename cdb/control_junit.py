#!/usr/bin/env python

from cdb.cdb_utils import control_junit

if __name__ == '__main__':
    control_junit()
