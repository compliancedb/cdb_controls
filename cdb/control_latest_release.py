#!/usr/bin/env python

from cdb.cdb_utils import control_latest_release

if __name__ == '__main__':
    control_latest_release()
