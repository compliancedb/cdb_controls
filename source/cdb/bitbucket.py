import json
import os
import sys

import requests

from cdb.cdb_utils import build_evidence_dict, send_evidence, set_artifact_sha_env_variable_from_file_or_image


def put_bitbucket_pull_request(project_config_file):

    is_compliant, pull_requests = get_pull_request_for_current_commit()

    print("Found pull request info: " + str(pull_requests))
    evidence = build_evidence_dict(
        is_compliant=is_compliant,
        evidence_type="pull_request",
        description="Bitbucket pull request",
        build_url=get_bitbucket_repo_url())
    evidence["contents"]["source"] = pull_requests
    send_evidence(project_config_file, evidence)


def get_pull_request_for_current_commit():
    workspace = os.getenv('BITBUCKET_WORKSPACE', None)
    repository = os.getenv('BITBUCKET_REPO_SLUG', None)
    commit = os.getenv('BITBUCKET_COMMIT', None)
    user = os.getenv('BITBUCKET_API_USER', None)
    password = os.getenv('BITBUCKET_API_TOKEN', None)
    force_compliant = os.getenv('CDB_FORCE_COMPLIANT', "FALSE") == "TRUE"
    fail_pipeline = os.getenv('CDB_FAIL_PIPELINE', "TRUE") == "TRUE"
    
    is_compliant, pull_requests = get_pull_requests_from_bitbucket_api(
        workspace=workspace, repository=repository, commit=commit, username=user, password=password, 
        force_compliant=force_compliant, fail_pipeline=fail_pipeline)
    return is_compliant, pull_requests


def verify_pull_request_evidence(pull_requests_evidence, force_compliant=False):
    if pull_requests_evidence == []:
        return False
    return True


def get_pull_requests_from_bitbucket_api(workspace, repository, commit, username, password, force_compliant, fail_pipeline):
    pull_requests_evidence = []
    is_compliant = False

    url = f"https://api.bitbucket.org/2.0/repositories/{workspace}/{repository}/commit/{commit}/pullrequests"
    print("Getting pull requests from " + url)
    r = requests.get(url, auth=(username, password))
    if r.status_code == 200:
        is_compliant = parse_response(commit, force_compliant, is_compliant, password, pull_requests_evidence, r,
                                      username)
    elif r.status_code == 202:
        print("Repository pull requests are still being indexed, please retry.")
        if fail_pipeline:
            sys.exit(1)
    elif r.status_code == 404:
        print("Repository does not exists or pull requests are not indexed. Please make sure Pull Request Commit Links app is installed")
        if fail_pipeline:
            sys.exit(2)
    else:
        print(f"Exception occurred in fetching pull requests. Http return code is {r.status_code}")
        print("    " + r.text)
        if fail_pipeline:
            sys.exit(3)

    return is_compliant, pull_requests_evidence


def parse_response(commit, force_compliant, is_compliant, password, pull_requests_evidence, r, username):
    print("Pull requests response: " + r.text)
    pull_requests_json = json.loads(r.text)
    pull_requests = pull_requests_json["values"]
    for pr in pull_requests:
        pr_evidence = {}

        pr_html_url = pr['links']['html']['href']
        pr_api_url = pr['links']['self']['href']

        pr_evidence['pullRequestMergeCommit'] = commit
        pr_evidence['pullRequestURL'] = pr_html_url

        pr_evidence = get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password)

        pull_requests_evidence.append(pr_evidence)
    is_compliant = verify_pull_request_evidence(pull_requests_evidence, force_compliant)
    return is_compliant


def get_pull_request_details_from_bitbucket(pr_evidence, pr_api_url, username, password):
    r = requests.get(pr_api_url, auth=(username, password))
    if r.status_code == 200:
        pr_json = json.loads(r.text)
        pr_evidence['pullRequestState'] = pr_json['state']
        participants = pr_json['participants']
        approvers = ""
        if len(participants) > 0:
            for participant in participants:
                if participant['approved']:
                    approvers = approvers + participant['user']['display_name'] + ","
            approvers = approvers[:-1]
            pr_evidence['approvers'] = approvers
        else:
            print("No approvers found")
    else:
        print("Error occurred in fetching pull request details. Please review repository permissions.")
    return pr_evidence


def get_bitbucket_repo_url():
    bb_workspace = os.environ.get("BITBUCKET_WORKSPACE")
    bb_repo_slug = os.environ.get("BITBUCKET_REPO_SLUG")
    repo_url = f"https://bitbucket.org/{bb_workspace}/{bb_repo_slug}"
    return repo_url


def adapt_put_artifact_image_env_variables():
    adapt_bitbucket_env_variables()
    set_artifact_sha_env_variable_from_file_or_image()


def adapt_create_release_variables():
    adapt_bitbucket_env_variables()
    set_artifact_sha_env_variable_from_file_or_image()
    os.environ["CDB_SRC_REPO_ROOT"] = os.environ.get("BITBUCKET_CLONE_DIR") + "/"


def adapt_create_deployment_variables():
    adapt_bitbucket_env_variables()


def adapt_control_junit_env_variables():
    adapt_bitbucket_env_variables()
    set_artifact_sha_env_variable_from_file_or_image()


def adapt_put_artifact_env_variables():
    adapt_bitbucket_env_variables()
    set_artifact_sha_env_variable_from_file_or_image()


def adapt_bitbucket_env_variables():
    repo_url = get_bitbucket_repo_url()

    bb_commit = os.environ.get("BITBUCKET_COMMIT")
    bb_build_number = os.environ.get("BITBUCKET_BUILD_NUMBER")

    build_url = f"{repo_url}/addon/pipelines/home#!/results/{bb_build_number}"

    os.environ["CDB_ARTIFACT_GIT_URL"] = f"{repo_url}/commits/{bb_commit}"
    os.environ["CDB_ARTIFACT_GIT_COMMIT"] = bb_commit
    os.environ["CDB_BUILD_NUMBER"] = bb_build_number
    os.environ["CDB_CI_BUILD_URL"] = build_url