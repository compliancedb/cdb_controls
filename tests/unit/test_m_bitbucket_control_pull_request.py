from cdb.bitbucket import put_bitbucket_pull_request
from commands import run

from tests.utils import *

APPROVAL_DIR = "tests/unit/approved_executions"
APPROVAL_FILE = "test_m_bitbucket_control_pull_request"

DOMAIN = "app.compliancedb.com"
API_TOKEN = "5199831f4ee3b79e7c5b7e0ebe75d67aa66e79d4"

BB = 'bitbucket.org'
ORG = 'acme'
REPO = 'road-runner'
COMMIT = "abc50c8a53f79974d615df335669b59fb56a4ed3"
BUILD_NUMBER = '1975'

PROTOCOL = "docker://"
IMAGE_NAME = "acme/road-runner:4.67"
SHA256 = "aacdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee"

DESCRIPTION = "branch coverage"
EVIDENCE_TYPE = "unit_test"


def test_bitbucket(capsys, mocker):
    env = {
        "CDB_HOST": f"https://{DOMAIN}",
        "CDB_API_TOKEN": API_TOKEN,
        "CDB_ARTIFACT_SHA": SHA256,

        "BITBUCKET_API_TOKEN": "6199831f4ee3b79e7c5b7e0ebe75d67aa66e79d4",
        "BITBUCKET_API_USER": "test_user",

        "BITBUCKET_WORKSPACE": ORG,
        "BITBUCKET_REPO_SLUG": REPO,
        "BITBUCKET_COMMIT": COMMIT,
    }

    with dry_run(env):
        mocker.patch('cdb.bitbucket.requests.get',
                     return_value=MockedAPIResponse(200, mocked_bitbucket_pull_requests_api_response))
        mocker.patch('cdb.bitbucket.json.loads', return_value=mocked_bitbucket_pull_requests_api_response)
        mocker.patch('cdb.cdb_utils.load_project_configuration', return_value=test_pipefile)
        put_bitbucket_pull_request("tests/integration/test-pipefile.json")
        mocker.stopall()

    verify_approval(capsys, ["out"])

    # extract data from approved cdb text file
    import inspect
    this_test = inspect.stack()[0].function
    approved = f"{APPROVAL_DIR}/{APPROVAL_FILE}.{this_test}.approved.txt"
    with open(approved) as file:
        old_approval = file.read()
    _old_blurb, old_method, old_payload, old_url = extract_blurb_method_payload_url(old_approval)

    expected_method = "Putting"
    expected_url = f"https://{DOMAIN}/api/v1/projects/{ORG}/{REPO}/artifacts/{SHA256}"
    expected_payload = {
        "contents": {
            "description": "Bitbucket pull request",
            "is_compliant": True,
            "source": [
                {
                    "approvers": "test_username",
                    "pullRequestMergeCommit": COMMIT,
                    "pullRequestState": "OPEN",
                    "pullRequestURL": "test_html_uri"
                }
            ],
            "url": f"https://{BB}/{ORG}/{REPO}"
        },
        "evidence_type": "pull_request"
    }

    # verify data from approved cdb text file
    assert old_method == expected_method
    assert old_url == expected_url
    assert old_payload == expected_payload

    # make merkely call
    #...




class MockedAPIResponse:
    def __init__(self, status_code, response_body=None):
        self.status_code = status_code
        self.response_body = response_body

    @property
    def text(self):
        return str(self.response_body)


mocked_bitbucket_pull_requests_api_response = {
    "values": [{
        "links": {
            "self": {"href": "test_self_uri", "name": "test_self"},
            "html": {"href": "test_html_uri", "name": "test_html"},
        },
        "id": "1",
        "title": "test pull request",
    }],
    "state": "OPEN",
    "participants": [
        {
            "approved": True,
            "state": "approved",
            "user": {
                "display_name": "test_username"
            }
        }
    ]
}


test_pipefile = {
    "owner": ORG,
    "name": REPO,
    "description": "Test Pipeline Controls for Merkely",
    "visibility": "public",
    "template": [
        "artifact",
        "unit_test",
        "coverage"
    ]
}


def new_control_pull_request_env():
    return {
        "MERKELY_COMMAND": "control_pull_request",
        "MERKELY_FINGERPRINT": f"{PROTOCOL}{IMAGE_NAME}",
        "MERKELY_API_TOKEN": API_TOKEN,
        "MERKELY_HOST": f"https://{DOMAIN}",
        "MERKELY_IS_COMPLIANT": "TRUE",
        "MERKELY_EVIDENCE_TYPE": EVIDENCE_TYPE,
        "MERKELY_DESCRIPTION": DESCRIPTION,

        "BITBUCKET_API_TOKEN": "6199831f4ee3b79e7c5b7e0ebe75d67aa66e79d4",
        "BITBUCKET_API_USER": "test_user",

        "BITBUCKET_WORKSPACE": ORG,
        "BITBUCKET_REPO_SLUG": REPO,
        "BITBUCKET_COMMIT": COMMIT,
        "BITBUCKET_BUILD_NUMBER": BUILD_NUMBER,
    }
