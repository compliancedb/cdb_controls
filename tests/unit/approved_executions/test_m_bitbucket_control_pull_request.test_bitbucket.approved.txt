Getting pull requests from https://api.bitbucket.org/2.0/repositories/acme/road-runner/commit/abc50c8a53f79974d615df335669b59fb56a4ed3/pullrequests
Pull requests response: {'values': [{'links': {'self': {'href': 'test_self_uri', 'name': 'test_self'}, 'html': {'href': 'test_html_uri', 'name': 'test_html'}}, 'id': '1', 'title': 'test pull request'}], 'state': 'OPEN', 'participants': [{'approved': True, 'state': 'approved', 'user': {'display_name': 'test_username'}}]}
Found pull request info: [{'pullRequestMergeCommit': 'abc50c8a53f79974d615df335669b59fb56a4ed3', 'pullRequestURL': 'test_html_uri', 'pullRequestState': 'OPEN', 'approvers': 'test_username'}]
{'evidence_type': 'pull_request', 'contents': {'is_compliant': True, 'url': 'https://bitbucket.org/acme/road-runner', 'description': 'Bitbucket pull request', 'source': [{'pullRequestMergeCommit': 'abc50c8a53f79974d615df335669b59fb56a4ed3', 'pullRequestURL': 'test_html_uri', 'pullRequestState': 'OPEN', 'approvers': 'test_username'}]}}
Docker image digest found from environment variable
IMAGE DIGEST: aacdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee
Putting this payload:
{
    "contents": {
        "description": "Bitbucket pull request",
        "is_compliant": true,
        "source": [
            {
                "approvers": "test_username",
                "pullRequestMergeCommit": "abc50c8a53f79974d615df335669b59fb56a4ed3",
                "pullRequestState": "OPEN",
                "pullRequestURL": "test_html_uri"
            }
        ],
        "url": "https://bitbucket.org/acme/road-runner"
    },
    "evidence_type": "pull_request"
}
To this url: https://app.compliancedb.com/api/v1/projects/acme/road-runner/artifacts/aacdaef69c676c2466571d3288880d559ccc2032b258fc5e73f99a103db462ee
DRY RUN: Put not sent
